REGRESSION (282524@main) (iOS): Tap and click events unresponsive after closing Safari and coming back

rr-codes opened the pull request:
#### 4f5b1337cdf766480f21401c40e8f954a970a8d6
<pre>
REGRESSION (282524@main) (iOS): Tap and click events unresponsive after closing Safari and coming back
<a href="https://bugs.webkit.org/show_bug.cgi?id=279146">https://bugs.webkit.org/show_bug.cgi?id=279146</a>
<a href="https://rdar.apple.com/135211157">rdar://135211157</a>

Reviewed by Abrar Rahman Protyasha.

In `PointerCaptureController`, there is logic which tries to not execute a mouse or pointer event
if there is a capturing event whose state is &quot;Ready&quot;. However, prior to 282524@main, this logic
only worked to not execute pointer events, but would still in fact execute mouse events.

After 282524@main, since click, auxclick, and contextmenu are now pointer events, the `PointerCaptureController` logic
is now successful in preventing both mouse _and_ pointer events.

This in itself isn&apos;t a problem, and is the desired behavior when there is an active capture state.

However, in some cases (such as dismissing Safari), a touch event ends up beginning but never getting
cancelled properly. Specifically, while a `pointercancel` event is dispatched, the associated capture
data&apos;s state was never properly updated.

Fix this by calling `cancelPointer`, which properly updates the state and also more closely adheres to
the pointer event specification with regards to which events should be dispatched.

Fixing this then in turn fixes the original issue, since now there will no erroneous states which are active.

A test will be added in a follow up.

* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
(WebCore::PointerCaptureController::cancelPointer):
* Source/WebCore/page/PointerCaptureController.h:

Canonical link: <a href="https://commits.webkit.org/283187@main">https://commits.webkit.org/283187@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a70ced822c53555503e69586d44eb4e359fce2af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69551 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16134 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67643 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16414 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11183 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68592 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56708 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33243 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38148 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14087 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15010 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71256 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9479 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9511 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56774 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14437 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7831 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41782 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -313,6 +313,11 @@ void PointerCaptureController::dispatchEventForTouchAtIndex(EventTarget& target,
     }
 
+    if (pointerEvent->type() == eventNames().pointercancelEvent) {
+        cancelPointer(pointerEvent->pointerId(), platformTouchEvent.touchLocationAtIndex(index), pointerEvent.ptr());
```
WPE will require something like `documentPointForWindowPoint(<frame>, platformTouchEvent.touchPoints()[index].pos())` I suppose

rr-codes replied:
In the interest of being able to merge this soon, I just added a platform guard for iOS family instead

rr-codes committed:
REGRESSION (282524@main) (iOS): Tap and click events unresponsive after closing Safari and coming back
https://bugs.webkit.org/show_bug.cgi?id=279146
rdar://135211157

Reviewed by Abrar Rahman Protyasha.

In `PointerCaptureController`, there is logic which tries to not execute a mouse or pointer event
if there is a capturing event whose state is "Ready". However, prior to 282524@main, this logic
only worked to not execute pointer events, but would still in fact execute mouse events.

After 282524@main, since click, auxclick, and contextmenu are now pointer events, the `PointerCaptureController` logic
is now successful in preventing both mouse _and_ pointer events.

This in itself isn't a problem, and is the desired behavior when there is an active capture state.

However, in some cases (such as dismissing Safari), a touch event ends up beginning but never getting
cancelled properly. Specifically, while a `pointercancel` event is dispatched, the associated capture
data's state was never properly updated.

Fix this by calling `cancelPointer`, which properly updates the state and also more closely adheres to
the pointer event specification with regards to which events should be dispatched.

Fixing this then in turn fixes the original issue, since now there will no erroneous states which are active.

A test will be added in a follow up.

* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
(WebCore::PointerCaptureController::cancelPointer):
* Source/WebCore/page/PointerCaptureController.h:

Canonical link: https://commits.webkit.org/283187@main
